### PR TITLE
DB-2006 Basic AwardeeOrRecipientUniqueIdentifier validations

### PR DESCRIPTION
**Pull request by michaeldhess**
Acceptance Criteria:

* AwardeeOrRecipientUniqueIdentifier
  * Not required for Loans, Direct Payments, Insurance, and Other types of financial assistance; this field is optional for these records, and if provided will not be validated other than checking that that it is 9 digits.
  * Field must be blank for aggregate records (i.e., when RecordType = 1) and individual recipients (i.e., when BusinessTypes includes "P").

(All credit to @nmonga91)

Technical Release Notes:
* New SQL rules

### CODE REVIEW

File: dataactvalidator/config/sqlrules/d31_detached_award_financial_assistance_2.sql [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    awardee_or_recipient_uniqu
+FROM detached_award_financial_assistance
+WHERE submission_id = {0}
+    AND COALESCE(assistance_type, '') IN ('02', '03', '04', '05')
+    AND (CASE
+            WHEN pg_temp.is_date(COALESCE(action_date, '0'))
+            THEN
+                CAST(action_date as DATE)
+        END) > CAST('10/01/2010' as DATE)
+    AND (awardee_or_recipient_uniqu IS NULL or awardee_or_recipient_uniqu = '')
```
Comment (alburde1) [inline dataactvalidator/config/sqlrules/d31_detached_award_financial_assistance_2.sql]: We use COALESCE for this same purpose in the above case. Can we choose one method for this PR and use it throughout? (doesn't really matter which)

File: tests/unit/dataactvalidator/test_d31_detached_award_financial_assistance_1.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    """ Test success for AwardeeOrRecipientUniqueIdentifier Field must be blank for aggregate records
+    (i.e., when RecordType = 1) and individual recipients. """
+    det_award_1 = DetachedAwardFinancialAssistanceFactory(record_type=1, business_types="ABP",
+                                                          awardee_or_recipient_uniqu='')
+    det_award_2 = DetachedAwardFinancialAssistanceFactory(record_type=1, business_types="ABC",
+                                                          awardee_or_recipient_uniqu='')
+    det_award_3 = DetachedAwardFinancialAssistanceFactory(record_type=2, business_types="ABP",
+                                                          awardee_or_recipient_uniqu='')
+    det_award_4 = DetachedAwardFinancialAssistanceFactory(record_type=2, business_types="ABC",
+                                                          awardee_or_recipient_uniqu='test')
```
Comment (alburde1) [inline tests/unit/dataactvalidator/test_d31_detached_award_financial_assistance_1.py]: Can we have a test or two with P in different locations and as a lowercase character?

File: tests/unit/dataactvalidator/test_d31_detached_award_financial_assistance_1.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    errors = number_of_errors(_FILE, database, models=[det_award_1, det_award_2, det_award_3, det_award_4])
+    assert errors == 0
+
+
+def test_failure(database):
+    """ Test failure for AwardeeOrRecipientUniqueIdentifier Field must be blank for aggregate records
+    (i.e., when RecordType = 1) and individual recipients. """
+
+    det_award = DetachedAwardFinancialAssistanceFactory(record_type=1, business_types="ABP",
+                                                        awardee_or_recipient_uniqu='test')
```
Comment (alburde1) [inline tests/unit/dataactvalidator/test_d31_detached_award_financial_assistance_1.py]: Same as above, can we include a test or two where P is in a different location and one that it's lowercase?

File: tests/unit/dataactvalidator/test_d31_detached_award_financial_assistance_2.py [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+def test_success(database):
+    """ Test success for AwardeeOrRecipientUniqueIdentifier is required for AssistanceType of 02, 03, 04, or 05 whose
+    ActionDate after October 1, 2010. """
+    det_award_1 = DetachedAwardFinancialAssistanceFactory(assistance_type="02", action_date="10/02/2010",
+                                                          awardee_or_recipient_uniqu='test')
+    det_award_2 = DetachedAwardFinancialAssistanceFactory(assistance_type="03", action_date="09/01/2010",
+                                                          awardee_or_recipient_uniqu='')
+    det_award_3 = DetachedAwardFinancialAssistanceFactory(assistance_type="06", action_date="10/02/2010",
+                                                          awardee_or_recipient_uniqu='')
```
Comment (alburde1) [inline tests/unit/dataactvalidator/test_d31_detached_award_financial_assistance_2.py]: Just to be thorough, it'd be nice to have tests for `03`, `04`, and `05`. It seems like overkill but it isn't many extras and in case something changes with the numbers we'll catch it

File: tests/unit/dataactvalidator/test_d31_detached_award_financial_assistance_1.py [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    det_award_5 = DetachedAwardFinancialAssistanceFactory(record_type=2, business_types="apc",
+                                                          awardee_or_recipient_uniqu='')
+    det_award_6 = DetachedAwardFinancialAssistanceFactory(record_type=2, business_types="APC",
+                                                          awardee_or_recipient_uniqu='')
+    det_award_7 = DetachedAwardFinancialAssistanceFactory(record_type=2, business_types="abp",
+                                                          awardee_or_recipient_uniqu='')
+    det_award_8 = DetachedAwardFinancialAssistanceFactory(record_type=2, business_types="ABP",
+                                                          awardee_or_recipient_uniqu='')
+    det_award_9 = DetachedAwardFinancialAssistanceFactory(record_type=2, business_types="ABC",
+                                                          awardee_or_recipient_uniqu='test')
```
Comment (alburde1) [inline tests/unit/dataactvalidator/test_d31_detached_award_financial_assistance_1.py]: You don't need to add any more tests, but just take one or two of these and add `None` instead of `''` for `awardee_or_recipient_uniqu`

File: tests/unit/dataactvalidator/test_d31_detached_award_financial_assistance_2.py [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    det_award_5 = DetachedAwardFinancialAssistanceFactory(assistance_type="02", action_date="09/01/2010",
+                                                          awardee_or_recipient_uniqu='')
+    det_award_6 = DetachedAwardFinancialAssistanceFactory(assistance_type="03", action_date="09/01/2010",
+                                                          awardee_or_recipient_uniqu='')
+    det_award_7 = DetachedAwardFinancialAssistanceFactory(assistance_type="04", action_date="09/01/2010",
+                                                          awardee_or_recipient_uniqu='')
+    det_award_8 = DetachedAwardFinancialAssistanceFactory(assistance_type="05", action_date="09/01/2010",
+                                                          awardee_or_recipient_uniqu='')
+    det_award_9 = DetachedAwardFinancialAssistanceFactory(assistance_type="06", action_date="10/02/2010",
+                                                          awardee_or_recipient_uniqu='')
```
Comment (alburde1) [inline tests/unit/dataactvalidator/test_d31_detached_award_financial_assistance_2.py]: One more test to show that non-blank `awardee_or_recipient_uniqu` still work in `09/01/2010`. Also, update one of the existing tests to use `None` instead of `''`

File: tests/unit/dataactvalidator/test_d31_detached_award_financial_assistance_2.py [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+    ActionDate after October 1, 2010. """
+
+    det_award_1 = DetachedAwardFinancialAssistanceFactory(assistance_type="02", action_date="10/02/2010",
+                                                          awardee_or_recipient_uniqu='')
+    det_award_2 = DetachedAwardFinancialAssistanceFactory(assistance_type="03", action_date="10/02/2010",
+                                                          awardee_or_recipient_uniqu='')
+    det_award_3 = DetachedAwardFinancialAssistanceFactory(assistance_type="04", action_date="10/02/2010",
+                                                          awardee_or_recipient_uniqu='')
+    det_award_4 = DetachedAwardFinancialAssistanceFactory(assistance_type="05", action_date="10/02/2010",
+                                                          awardee_or_recipient_uniqu='')
```
Comment (alburde1) [inline tests/unit/dataactvalidator/test_d31_detached_award_financial_assistance_2.py]: Update one of these to be `None` as well
